src/devices/Arduino/ArduinoBoard.cs

```diff
@@ -0,0 +1,177 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Device.Analog;
+using System.Text;
+using System.Device.Gpio;
+using System.Device.I2c;
+using System.Device.Pwm;
+using System.Device.Spi;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using Iot.Device.Spi;
+
+#pragma warning disable CS1591
+
+namespace Iot.Device.Arduino
+{
+    /// <summary>
+    /// Implements an interface to an arduino board which is running Firmata.
+    /// See documentation on how to prepare your arduino board to work with this.
+    /// Note that the program will run on the PC, so you cannot disconnect the
+    /// Arduino while this driver is connected.
+    /// </summary>
+    public class ArduinoBoard : IDisposable
+    {
+        private Stream _serialPortStream;
+        private FirmataDevice _firmata;
+        private Version _firmwareVersion;
+        private string _firmwareName;
+        private List<SupportedPinConfiguration> _supportedPinConfigurations;
+
+        public ArduinoBoard(Stream serialPortStream)
+        {
+            _serialPortStream = serialPortStream;
+        }
+
+        public event Action<string, Exception> LogMessages;
+
+        public virtual void Initialize()
```

Comment:
NIT: we usually call Initialize from constructor and make it private.

Reply:
I know, but I'm not sure this is a good idea here, since this eventually needs to do a lot of stuff and can fail for various reasons. And I don't like throwing complex exceptions from constructors. 
This has several reasons:
- If an error happens, you have to handle them immediately and free resources in a finaly clause (i.e. if you open a serial port and then detect that the device is not there)
- It makes unit testing very hard, because you can't easily introduce Mocks.
- There's no way the client can attempt to continue after an error (i.e. the initialization fails, because a wrong firmware is on the device, then the client may go to error state but still have a device connection to maybe upload the correct fw).

Reply:
That makes sense and I agree with you it’s best to keep it separate from the constructor. Can we then have a private field `bool _initialized` that we set to false until we call this method, and then make it so that every operation that requires initialization to first check if we need to call initialize and if so call it? What do you think about that?

Reply:
Yes, that sounds fine.